javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -133,12 +135,13 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
+        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
         Supplier<? extends T> tmp = supplier;
         if (tmp != null) {
             synchronized (this) {
                 tmp = supplier;
                 if (tmp != null) {
-                    value = supplier.get();
+                    value = tmp.get();
```

Comment:
Reads of volatile variables need to access the main memory to ensure the value is actual. `tmp` could be named `cachedSupplier` or `localSupplier`.

Because we return `value` instead of `supplier`, we do not need to cache the `supplier` before entering the synchronized block.

Within the synchronized block we have two reads (1. not-null check, 2. get() call). So a changed supplier saves one main-memory read.

The optimized code then looks like this:

```java
public T get() {
    if (this.supplier != null) {
        synchronized (this) {
            // only one volatile read from main-memory
            Supplier<? extends T> supplier = this.supplier;
            if (supplier != null) {
                value = supplier.get();
                this.supplier = null; // free mem
            }
        }
    }
    return value;
}
```